config/validate/validate.go

```diff
@@ -0,0 +1,174 @@
+	return line, col, highlight
+}
+
+func Validate(cfg types.Config, ast json.Node, source io.ReadSeeker) report.Report {
+	v := reflect.ValueOf(cfg)
+	report := validate(v, ast, source)
+	return report
+}
+
+// Validate walks down a struct tree calling AssertValid on every node that implements it, building
+// A report of all the errors, warnings, info, and deprecations it encounters
+func validate(vObj reflect.Value, ast json.Node, source io.ReadSeeker) (r report.Report) {
+	if !vObj.IsValid() {
+		return
+	}
+
+	off := ast.End
+
+	if obj, ok := canCallAssertValid(vObj); ok {
+		if err := obj.AssertValid(); err != nil {
+			line, col, highlight := posFromOffset(off, source)
+			r.Add(report.Entry{
+				Kind:      report.EntryError,
+				Message:   err.Error(),
+				Line:      line,
+				Column:    col,
+				Highlight: highlight,
+			})
+			return //dont recurse on invalid inner nodes, it mostly leads to bogus messages
+		}
+	}
+
+	switch vObj.Kind() {
+	case reflect.Ptr:
+		sub_report := validate(vObj.Elem(), ast, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	case reflect.Struct:
+		sub_report := validateStruct(vObj, ast, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	case reflect.Slice:
+		for i := 0; i < vObj.Len(); i++ {
+			sub_node := ast
+			if val, ok := ast.Value.([]json.Node); ok {
+				sub_node = val[i]
+			}
+			sub_report := validate(vObj.Index(i), sub_node, source)
+			sub_report.AddLines(off)
+			r.Merge(sub_report)
+		}
+	}
+	return
+}
+
+func ValidateWithoutSource(cfg types.Config) (report report.Report) {
+	return Validate(cfg, json.Node{}, nil)
+}
+
+func validateStruct(vObj reflect.Value, ast json.Node, source io.ReadSeeker) report.Report {
+	off := ast.Start
+	r := report.Report{}
+	keys, ok := ast.Value.(map[string]json.Node)
+	usedKeys := map[string]bool{} // Maintain a set of key's that have been used
+	tags := []string{}            // Maintain a list of all the tags in the struct for fuzzy matching later
+
+	for i := 0; i < vObj.Type().NumField(); i++ {
+		// Default to deepest node if the node's type isn't an object,
+		// such as when a json string actually unmarshal to structs (like with version)
+		sub_node := ast
+		tag := strings.Split(vObj.Type().Field(i).Tag.Get("json"), ",")[0]
+		tags = append(tags, tag) //build list of json key names
+		if ok {
+			sub_node = keys[tag]
+			usedKeys[tag] = true
+		}
+
+		sub_report := validate(vObj.Field(i), sub_node, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	}
+	if !ok {
+		// If this node's type isn't an json object, there can't be unused keys.
+		return r
+	}
+
+	for k, v := range keys {
+		if usedKeys[k] {
+			continue
+		}
+		line, col, highlight := posFromOffset(v.KeyEnd, source)
+		typo := similar(k, tags)
+
+		message := fmt.Sprintf("Config has unused key: %s", k)
+		if typo != "" {
+			message = fmt.Sprintf("Config has unused key: %s. Did you mean \"%s\"?", k, typo)
```

Comment:
I was thinking we make "Config has unrecognized key: %s" a warning and then provide hinting via an info on the same line.